### ISSUE TITLE
[SPIR-V] dot product using half3 causes compiler to crash

### ISSUE THREAD

**Issue by jmccarthy634** [microsoft/DirectXShaderCompiler#5048]
Attempting to compile the hollowing HLSL to spir-v with the arguments `dxc -spirv -fspv-target-env=vulkan1.1 -fvk-use-dx-layout -O3 -T cs_6_0 -E csMain E:\dotproduct.hlsl` causes the compiler to crash:

```
RWTexture2D<float> g_output : register(u0);
Texture2D<float4> g_input : register(t0);

half3 RGBToYCbCr_half(half3 rgb)
{
 return half3(dot(rgb, half3(0.2126, 0.7152, 0.0722)),
  dot(rgb, half3(-0.114572, -0.385428, 0.5)),
  dot(rgb, half3(0.5, -0.454153, -0.0458471)));
}

[numthreads(8, 8, 1)]
void csMain(
 int2 dispatchThreadId : SV_DispatchThreadID)
{
  float4 input = g_input.Load(int3(dispatchThreadId, 0));
  half3 workingColor = RGBToYCbCr_half(half3(input.xyz));

  g_output[dispatchThreadId] = float(workingColor.x);
}
```

Callstack:
```
Callstack:
    KERNELBASE.dll(0x7ffa9481cd29): ???
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(10238): clang::spirv::SpirvEmitter::processIntrinsicDot
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(8228): clang::spirv::SpirvEmitter::processIntrinsicCallExpr
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(2691): clang::spirv::SpirvEmitter::doCallExpr
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1096): clang::spirv::SpirvEmitter::doExpr
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1148): clang::spirv::SpirvEmitter::loadIfGLValue
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(3028): clang::spirv::SpirvEmitter::doCastExpr
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1067): clang::spirv::SpirvEmitter::doExpr
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(6522): clang::spirv::SpirvEmitter::processBinaryOp
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(2659): clang::spirv::SpirvEmitter::doBinaryOperator
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1088): clang::spirv::SpirvEmitter::doExpr
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(2455): clang::spirv::SpirvEmitter::doIfStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1008): clang::spirv::SpirvEmitter::doStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1030): clang::spirv::SpirvEmitter::doStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1000): clang::spirv::SpirvEmitter::doStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(2496): clang::spirv::SpirvEmitter::doIfStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1008): clang::spirv::SpirvEmitter::doStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1000): clang::spirv::SpirvEmitter::doStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1000): clang::spirv::SpirvEmitter::doStmt
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(1443): clang::spirv::SpirvEmitter::doFunctionDecl
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(923): clang::spirv::SpirvEmitter::doDecl
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\SPIRV\SpirvEmitter.cpp(787): clang::spirv::SpirvEmitter::HandleTranslationUnit
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\Parse\ParseAST.cpp(166): clang::ParseAST
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\Frontend\FrontendAction.cpp(556): clang::ASTFrontendAction::ExecuteAction
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\lib\Frontend\FrontendAction.cpp(469): clang::FrontendAction::Execute
    E:\dev\dxcpackage\External\DirectXShaderCompiler\tools\clang\tools\dxcompiler\dxcompilerobj.cpp(1041): DxcCompiler::Compile
 ```
 Reproducible on a build from the (as of now) latest commit, 740dd0951. Does not occur on earlier build from 7a45cbbf8 that I have to hand, so this appears to be a regression.

**Comment by jmccarthy634** [comment]
Compiling with `-enable-16bit-types` makes the issue go away, so it looks like it really is an issue with internal conversion to and from half precision types within the compiler.

**Comment by Keenuts** [comment]
Hello! Thanks for the report.

Short repro:
```hlsl
[numthreads(1, 1, 1)]
void main()
{
  dot(half3(0, 0, 0), half3(0, 1, 2));
}
```

`dxc sample.hlsl -E main -spirv -T cs_6_6`

The issue is when building the `half dot(half3, half3)`, we check that the return type of the `dot` function is the same as the input params. But here is the AST:
```
CallExpr 0x555555a61500 'half'
|-ImplicitCastExpr 0x555555a614e8 'half (*)(vector<float, 3>, vector<float, 3>)' <FunctionToPointerDecay>
| `-DeclRefExpr 0x555555a61450 'half (vector<float, 3>, vector<float, 3>)' lvalue Function 0x555555a61238 'dot' 'half (vector<float, 3>, vector<float, 3>)'
|-ImplicitCastExpr 0x555555a61538 'vector<float, 3>' <HLSLCC_FloatingCast>
| `-CXXFunctionalCastExpr 0x555555a61040 'half3':'vector<half, 3>' functional cast to half3 <NoOp>
```

16-bit types not being enabled, they are using 32-bit floats. So the dot params are float.
But the dot return value is still a `half`.
When enabling 16-bit types, the `dot` prototype is correctly `half(half3, half3)`.
Not sure why.

**Comment by s-perron** [comment]
I looked at the original test case, and it seem like the only problem is that this assert it being triggered.

The assert checks that the return type on the `dot` is the same as the component type of the parameters. It fails in this case, because the spirv emitter change the `half` types to `float` as it goes along. In this case, the parameters have already been changed to float, but the return type is still half. If we let it go, the return type will be changed to float.

This seems to be legit, so I will simply remove the assert.

Note that I still see the assert being triggered with commit 7a45cbb. I wonder if the original poster was testing a release build with the older commit, which is why it worked.
